sympy/combinatorics/free_group.py

```diff
@@ -0,0 +1,916 @@
+
+    @property
+    def rank(self):
+        r"""
+        In group theory, the `rank` of a group `G`, denoted `G.rank`,
+        can refer to the smallest cardinality of a generating set
+        for G, that is
+
+        \operatorname{rank}(G)=\min\{ |X|: X\subseteq G, \langle X\rangle =G\}.
+
+        """
+        return self._rank
+
+    @property
+    def is_abelian(self):
+        """Test if the group is Abelian.
+
+        Examples
+        ========
+
+        >>> from sympy import FreeGroup
+        >>> f = FreeGroup( 4 )
+        >>> f.is_abelian
+        False
+
+        >>> g = FreeGroup( 0 )
+        >>> g.is_abelian
+        True
+
+        """
+        if self.rank == 0 or self.rank == 1:
+            return True
+        else:
+            return False
+
+    def contains(self, g):
+        """Test if Free Group element ``g`` belong to self, ``G``.
+
+        In mathematical terms any linear combination of generators
+        of a Free Group is contained in it.
+
+        Examples
+        ========
+
+        >>> from sympy import FreeGroup
+        >>> f = FreeGroup( 4 )
+        >>> f.contains(f[0]**3*f[1]**2)
+        True
+
+        """
+        if not isinstance(i, FreeGroupElm):
+            return False
+        elif self != i.group:
+            return False
+        else:
+            return True
+
+    def is_subgroup(self, F):
+        """Return True if all elements of `self` belong to `F`.
+
+        Examples
+        ========
+        """
+        return F.is_group and all([self.contains(gen) for gen in F.generators])
+
+    def assign_variables(self):
+        """
+        If  self is a group, whose generators are represented by symbols
+        (for example a free group, a finitely presented group or a pc group)
+        this function assigns these generators to global variables with the
+        same names.
+        """
+        pass
+
+
+############################################################################
+#                          FreeGroupElm                                    #
+############################################################################
+
+
+class FreeGroupElm(Basic):
+    """
+    ``FreeGroupElm`` is actually not usable as public import, since it is
+    always associated with a ``FreeGroup``. It is always called only by the
+    ``FreeGroup`` class. Called with a ``FreeGroup`` as the first argument and
+    the second argument called the `array_form` for the ``FreeGroupElm``, and
+    the third argument is a string representation for the element.
+
+    """
+    is_Identity = None
+    is_AssocWord = True
+
+    def __new__(cls, free_group, array_form, str_expr=""):
+        obj = Basic.__new__(cls, free_group, array_form, str_expr)
+
+        # `array_form` is used internally by the methods
+        # of `FreeGroupElm`
+        obj._group = free_group
+        obj._array_form = array_form
+        obj._str_expr = str_expr
+        return obj
+
+    @property
+    def group(self):
```

Comment:
This attribute of `group` will vary according to the Free Group element, each `FreeGroupElm` will have a different Free Group, which will be provided to `__new__(cls, free_group, ... )` just above this code.